File: hooks/test_vendored_skills.py
#!/usr/bin/env python3
"""Tests that the superpowers discipline skills are vendored self-contained into Genesis.

Genesis's /spec-forge workflow (skills spec-forge, forge-dev-agent, forge-review-agent) invokes
discipline skills that originate in the separate `superpowers` plugin. Those skills are vendored
into skills/ so a fresh Genesis install works without superpowers installed. This test proves the
vendoring is complete and dangling-reference-free:

  * ZERO `superpowers:` skill references remain anywhere under skills/ (all rewired to Genesis's
    bare-name convention).
  * Every skill the spec-forge / forge-* workflow references now exists under skills/<name>/SKILL.md.
  * The full transitive closure is present (11 skills) and every `../<sibling>` cross-skill path
    reference inside a vendored skill resolves to an existing sibling skill (no dangling refs).
  * Third-party license/attribution is preserved (VENDORED-superpowers-LICENSE + NOTICE.md entry).

Run:  python3 hooks/test_vendored_skills.py
"""
import os
import re
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(HERE)                       # hooks/ -> plugin root
SKILLS = os.path.join(REPO, "skills")

# The complete transitive closure vendored from superpowers 6.1.1.
VENDORED = [
    "brainstorming",
    "writing-plans",
    "using-git-worktrees",
    "test-driven-development",
    "systematic-debugging",
    "verification-before-completion",
    "requesting-code-review",
    "receiving-code-review",
    "finishing-a-development-branch",
    "subagent-driven-development",
    "executing-plans",
]

# Genesis workflow skills that drove the dependency (they reference the vendored discipline skills).
REFERENCING_DIRS = [
    "spec-forge",
    "forge-dev-agent",
    "forge-docs-agent",
    "forge-review-agent",
    "forge-spec-agent",
    "forge-verify-agent",
]


def skill_dirs():
    return {d for d in os.listdir(SKILLS) if os.path.isdir(os.path.join(SKILLS, d))}


def all_skill_files():
    for root, _dirs, files in os.walk(SKILLS):
        for f in files:
            yield os.path.join(root, f)


def read(path):
    with open(path, encoding="utf-8", errors="replace") as fh:
        return fh.read()


def main():
    passed = failed = 0

    def check(name, cond):
        nonlocal passed, failed
        passed += 1 if cond else 0
        failed += 0 if cond else 1
        print(f"  {'PASS' if cond else 'FAIL'}  {name}")

    dirs = skill_dirs()

    # ---- 1. ZERO `superpowers:` references anywhere under skills/ ----
    offenders = []
    for path in all_skill_files():
        try:
            if "superpowers:" in read(path):
                offenders.append(os.path.relpath(path, REPO))
        except (OSError, UnicodeDecodeError):
            pass
    check(f"no `superpowers:` refs remain under skills/ (found in: {offenders})", not offenders)

    # ---- 2. Every vendored skill exists as skills/<name>/SKILL.md ----
    for name in VENDORED:
        check(f"vendored skill present: {name}/SKILL.md",
              os.path.isfile(os.path.join(SKILLS, name, "SKILL.md")))

    # ---- 3. Every skill referenced by spec-forge / forge-* exists under skills/ ----
    # Collect backtick-quoted skill-like tokens from the referencing files, restricted to the
    # vendored discipline-skill vocabulary, and assert each resolves to a real skill directory.
    vendored_set = set(VENDORED)
    referenced = set()
    tok = re.compile(r"`([a-z][a-z0-9-]+)`")
    for d in REFERENCING_DIRS:
        ddir = os.path.join(SKILLS, d)
        if not os.path.isdir(ddir):
            continue
        for root, _sub, files in os.walk(ddir):
            for f in files:
                if not f.endswith(".md"):
                    continue
                for m in tok.findall(read(os.path.join(root, f))):
                    if m in vendored_set:
                        referenced.add(m)
    check("spec-forge/forge-* reference at least the core discipline skills",
          {"test-driven-development", "verification-before-completion"} <= referenced)
    missing_referenced = sorted(n for n in referenced if n not in dirs)
    check(f"every discipline skill referenced by spec-forge/forge-* exists (missing: {missing_referenced})",
          not missing_referenced)

    # ---- 4. Transitive-closure integrity: every `../<sibling>` ref inside a vendored skill resolves ----
    rel = re.compile(r"\.\./([A-Za-z0-9_-]+)/")
    dangling = []
    # Skill names that a vendored skill may legitimately point at via ../ (sibling skills).
    for name in VENDORED:
        ndir = os.path.join(SKILLS, name)
        for root, _sub, files in os.walk(ndir):
            for f in files:
                if not (f.endswith(".md") or f.endswith(".sh")):
                    continue
                text = read(os.path.join(root, f))
                for target in rel.findall(text):
                    # ../<target>/ is a cross-skill sibling reference iff <target> names a skill dir
                    # under skills/. If <target> looks like a skill but is absent, it dangles.
                    if target in ("scripts", "references"):
                        continue
                    candidate = os.path.join(SKILLS, target)
                    if target not in dirs and not os.path.exists(candidate):
                        # Only flag ones that look like skill references (contain a hyphen or are known)
                        if "-" in target:
                            dangling.append(f"{name}: ../{target}/")
    check(f"no dangling ../<sibling> skill refs inside vendored skills (dangling: {dangling})",
          not dangling)

    # ---- 5. License / attribution preserved ----
    lic_path = os.path.join(SKILLS, "VENDORED-superpowers-LICENSE")
    lic = read(lic_path) if os.path.isfile(lic_path) else ""
    check("skills/VENDORED-superpowers-LICENSE exists with MIT + copyright holder",
          "MIT License" in lic and "Jesse Vincent" in lic)
    notice = read(os.path.join(REPO, "NOTICE.md"))
    check("NOTICE.md attributes superpowers (name + copyright + source URL)",
          "superpowers" in notice
          and "Jesse Vincent" in notice
          and "github.com/obra/superpowers" in notice)

    print(f"\n{passed} passed, {failed} failed")
    sys.exit(1 if failed else 0)


if __name__ == "__main__":
    main()

File: hooks/test_vendored_skills.js
#!/usr/bin/env node
/* Node tests that the superpowers discipline skills are vendored self-contained into Genesis.

   Faithful port of test_vendored_skills.py. Pure filesystem inspection (no subprocess, no logs). Proves the
   vendoring is complete and dangling-reference-free:
     * ZERO `superpowers:` skill references remain anywhere under skills/.
     * Every skill the spec-forge / forge-* workflow references now exists under skills/<name>/SKILL.md.
     * The full transitive closure (11 skills) is present and every `../<sibling>` cross-skill path resolves.
     * Third-party license/attribution is preserved.
   Mirrors the Python contract + case count (17 cases).

   Run:  node hooks/test_vendored_skills.js
*/
"use strict";
const fs = require("fs");
const path = require("path");

const HERE = __dirname;
const REPO = path.dirname(HERE); // hooks/ -> plugin root
const SKILLS = path.join(REPO, "skills");

// The complete transitive closure vendored from superpowers 6.1.1.
const VENDORED = [
  "brainstorming",
  "writing-plans",
  "using-git-worktrees",
  "test-driven-development",
  "systematic-debugging",
  "verification-before-completion",
  "requesting-code-review",
  "receiving-code-review",
  "finishing-a-development-branch",
  "subagent-driven-development",
  "executing-plans",
];

// Genesis workflow skills that drove the dependency (they reference the vendored discipline skills).
const REFERENCING_DIRS = [
  "spec-forge",
  "forge-dev-agent",
  "forge-docs-agent",
  "forge-review-agent",
  "forge-spec-agent",
  "forge-verify-agent",
];

function isDir(p) { try { return fs.statSync(p).isDirectory(); } catch (e) { return false; } }
function isFile(p) { try { return fs.statSync(p).isFile(); } catch (e) { return false; } }

function skill_dirs() {
  const out = new Set();
  let entries;
  try { entries = fs.readdirSync(SKILLS, { withFileTypes: true }); } catch (e) { return out; }
  for (const e of entries) if (e.isDirectory()) out.add(e.name);
  return out;
}

function walkFiles(dir) {
  const out = [];
  let entries;
  try { entries = fs.readdirSync(dir, { withFileTypes: true }); } catch (e) { return out; }
  for (const e of entries) {
    const full = path.join(dir, e.name);
    if (e.isDirectory()) out.push.apply(out, walkFiles(full));
    else out.push(full);
  }
  return out;
}

function read(p) {
  try { return fs.readFileSync(p, "utf8"); } catch (e) { return ""; }
}

function main() {
  let passed = 0, failed = 0;
  function check(name, cond) {
    if (cond) { passed += 1; } else { failed += 1; }
    console.log("  " + (cond ? "PASS" : "FAIL") + "  " + name);
  }

  const dirs = skill_dirs();

  // ---- 1. ZERO `superpowers:` references anywhere under skills/ ----
  const offenders = [];
  for (const p of walkFiles(SKILLS)) {
    if (read(p).indexOf("superpowers:") !== -1) offenders.push(path.relative(REPO, p));
  }
  check("no `superpowers:` refs remain under skills/ (found in: " + JSON.stringify(offenders) + ")", offenders.length === 0);

  // ---- 2. Every vendored skill exists as skills/<name>/SKILL.md ----
  for (const name of VENDORED) {
    check("vendored skill present: " + name + "/SKILL.md",
          isFile(path.join(SKILLS, name, "SKILL.md")));
  }

  // ---- 3. Every skill referenced by spec-forge / forge-* exists under skills/ ----
  const vendored_set = new Set(VENDORED);
  const referenced = new Set();
  const tok = /`([a-z][a-z0-9-]+)`/g;
  for (const d of REFERENCING_DIRS) {
    const ddir = path.join(SKILLS, d);
    if (!isDir(ddir)) continue;
    for (const f of walkFiles(ddir)) {
      if (!f.endsWith(".md")) continue;
      const text = read(f);
      let m;
      tok.lastIndex = 0;
      while ((m = tok.exec(text)) !== null) {
        if (vendored_set.has(m[1])) referenced.add(m[1]);
      }
    }
  }
  check("spec-forge/forge-* reference at least the core discipline skills",
        ["test-driven-development", "verification-before-completion"].every((n) => referenced.has(n)));
  const missing_referenced = Array.from(referenced).filter((n) => !dirs.has(n)).sort();
  check("every discipline skill referenced by spec-forge/forge-* exists (missing: " + JSON.stringify(missing_referenced) + ")",
        missing_referenced.length === 0);

  // ---- 4. Transitive-closure integrity: every `../<sibling>` ref inside a vendored skill resolves ----
  const rel = /\.\.\/([A-Za-z0-9_-]+)\//g;
  const dangling = [];
  for (const name of VENDORED) {
    const ndir = path.join(SKILLS, name);
    for (const f of walkFiles(ndir)) {
      if (!(f.endsWith(".md") || f.endsWith(".sh"))) continue;
      const text = read(f);
      let m;
      rel.lastIndex = 0;
      while ((m = rel.exec(text)) !== null) {
        const target = m[1];
        if (target === "scripts" || target === "references") continue;
        const candidate = path.join(SKILLS, target);
        if (!dirs.has(target) && !fs.existsSync(candidate)) {
          if (target.indexOf("-") !== -1) dangling.push(name + ": ../" + target + "/");
        }
      }
    }
  }
  check("no dangling ../<sibling> skill refs inside vendored skills (dangling: " + JSON.stringify(dangling) + ")",
        dangling.length === 0);

  // ---- 5. License / attribution preserved ----
  const lic_path = path.join(SKILLS, "VENDORED-superpowers-LICENSE");
  const lic = isFile(lic_path) ? read(lic_path) : "";
  check("skills/VENDORED-superpowers-LICENSE exists with MIT + copyright holder",
        lic.indexOf("MIT License") !== -1 && lic.indexOf("Jesse Vincent") !== -1);
  const notice = read(path.join(REPO, "NOTICE.md"));
  check("NOTICE.md attributes superpowers (name + copyright + source URL)",
        notice.indexOf("superpowers") !== -1
        && notice.indexOf("Jesse Vincent") !== -1
        && notice.indexOf("github.com/obra/superpowers") !== -1);

  console.log("\n" + passed + " passed, " + failed + " failed");
  process.exit(failed ? 1 : 0);
}

if (require.main === module) {
  main();
}
